shared.go

```diff
@@ -162,11 +165,34 @@ type Header struct {
 	Algorithm  string
 	Nonce      string
 
-	// Any headers not recognised above get unmarshaled from JSON in a generic
-	// manner and placed in this map.
+	// Unverified certificate chain parsed from x5c header.
+	certificates []*x509.Certificate
+
+	// Any headers not recognised above get unmarshaled
+	// from JSON in a generic manner and placed in this map.
 	ExtraHeaders map[HeaderKey]interface{}
 }
 
+// Certificates verifies & returns the certificate chain present
+// in the x5c header field of a message, if one was present. Returns
+// an error if there was no x5c header present or the chain could
+// not be validated with the given verify options.
+func (h Header) Certificates(opts x509.VerifyOptions) ([][]*x509.Certificate, error) {
+	if len(h.certificates) == 0 {
+		return nil, errors.New("square/go-jose: no x5c header present in message")
+	}
+
+	leaf := h.certificates[0]
+	if opts.Intermediates == nil {
+		opts.Intermediates = x509.NewCertPool()
+		for _, intermediate := range h.certificates[1:] {
```

Comment:
is there definitely >=1 intermediate? otherwise this might panic at runtime

Reply:
Excellent point I should fix that, there could definitely be zero here

Reply:
Oh, actually it's fine: there's at least one element and `c[1:]` will just be an empty slice:
```
❯❯❯ gore
gore version 0.2.6  :help for help
gore> c := []string{"test"}
[]string{
  "test",
}
gore> c[1:]
[]string{}
```

Reply:
Added some extra test cases to cover this